rust/src/http2/http2.rs

```diff
@@ -0,0 +1,726 @@
+                            match parser::http2_parse_frame_headers(rem, head.flags) {
+                                Ok((_, hs)) => {
+                                    tx.type_data = Some(HTTP2FrameTypeData::HEADERS(hs));
+                                }
+                                Err(nom::Err::Incomplete(_)) => {
+                                    return AppLayerResult::incomplete(
+                                        (il - input.len()) as u32,
+                                        (HTTP2_FRAME_HEADER_LEN + 1) as u32,
+                                    );
+                                }
+                                Err(_) => {
+                                    self.set_event(HTTP2Event::InvalidFrameData);
+                                }
+                            }
+                        }
+                        //ignore ping case with opaque u64
+                        _ => {}
+                    }
+                    self.transactions.push(tx);
+
+                    let hl = head.length as usize;
+                    if rem.len() < hl {
+                        let rl = rem.len() as u32;
+                        self.request_frame_size = head.length - rl;
+                        return AppLayerResult::ok();
+                    } else {
+                        input = &rem[hl..];
+                    }
+                }
+                Err(nom::Err::Incomplete(_)) => {
+                    //we may have consumed data from previous records
+                    if input.len() < HTTP2_FRAME_HEADER_LEN {
+                        return AppLayerResult::incomplete(
+                            (il - input.len()) as u32,
+                            HTTP2_FRAME_HEADER_LEN as u32,
+                        );
+                    } else {
+                        panic!("HTTP2 invalid length frame header");
+                    }
+                }
+                Err(_) => {
+                    self.set_event(HTTP2Event::InvalidFrameHeader);
+                    return AppLayerResult::err();
+                }
+            }
+        }
+        return AppLayerResult::ok();
+    }
+
+    fn parse_tc(&mut self, mut input: &[u8]) -> AppLayerResult {
+        //first consume frame bytes
+        let il = input.len();
+        if self.response_frame_size > 0 {
+            let ilen = input.len() as u32;
+            if self.response_frame_size >= ilen {
+                self.response_frame_size -= ilen;
+                return AppLayerResult::ok();
+            } else {
+                let start = self.response_frame_size as usize;
+                input = &input[start..];
+                self.response_frame_size = 0;
+            }
+        }
+        //then parse all we can
+        while input.len() > 0 {
+            match parser::http2_parse_frame_header(input) {
+                Ok((rem, head)) => {
+                    let mut tx = self.new_tx();
```

Comment:
the goal of a tx should be to match a response to a request, so not create txs for responses, but look up the request side tx.

Reply:
I know :-) this comes later in my TODO list.
I think I am getting a sense of how it should work with HTTP2